File: sf-ui/src/utils/links.js
// The same URL detection the API applies when it stores a post (see the Go
// utils.FirstURL). It is duplicated rather than fetched so the composer can
// preview the embed as the author types; the API's copy is the authority, and
// the two patterns are kept identical on purpose.
const URL_PATTERN = /https?:\/\/[^\s<>"'`]+[^\s<>"'`.,;:!?)\]}]/;
const BARE_URL_PATTERN = /^\s*(https?:\/\/\S+)\s*$/;

/** The first http(s) URL in text, or "" when there is none. */
export function firstUrl(text) {
  const found = (text || "").match(URL_PATTERN);
  if (found) return found[0];
  // A URL that is the whole text has no trailing character for the pattern's
  // final class to match.
  const bare = (text || "").match(BARE_URL_PATTERN);
  return bare ? bare[1] : "";
}

export default firstUrl;

File: sf-api/internal/utils/links.go
package utils

import "regexp"

// urlPattern matches an http(s) URL inside free text.
//
// The trailing class deliberately excludes the punctuation a sentence ends
// with, so "watch this: https://example.org/clip." doesn't capture the full
// stop, and a URL wrapped in brackets or quotes stops at the closing one. It
// is not a general-purpose URL grammar - it only has to recognise what
// somebody pastes into a post.
var urlPattern = regexp.MustCompile(`https?://[^\s<>"'\x60]+[^\s<>"'\x60.,;:!?)\]}]`)

// FirstURL returns the first http(s) URL in text, or "" if there is none.
//
// This is what replaced the separate "add a link" field: a post carries one
// piece of text, and whatever the author pasted into it is what gets rendered
// as a player or a link card.
func FirstURL(text string) string {
	found := urlPattern.FindString(text)
	if found == "" {
		// A URL that is the entire text has no trailing character for the
		// pattern's final class to match, so a bare "https://x.org" would
		// otherwise be missed.
		if bare := regexp.MustCompile(`^\s*(https?://\S+)\s*$`).FindStringSubmatch(text); bare != nil {
			return bare[1]
		}
	}
	return found
}
